rmw_connext_dynamic_cpp/src/templates.hpp

```diff
@@ -0,0 +1,1554 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TEMPLATES_HPP_
+#define TEMPLATES_HPP_
+
+#include <limits>
+#include <string>
+#include <vector>
+
+#include "rosidl_generator_cpp/message_type_support.hpp"
+
+#include "rosidl_generator_c/primitives_array_functions.h"
+#include "rosidl_generator_c/string.h"
+#include "rosidl_generator_c/string_functions.h"
+
+#include "rmw/impl/cpp/macros.hpp"
+
+#include "rosidl_typesupport_introspection_cpp/field_types.hpp"
+#include "rosidl_typesupport_introspection_cpp/identifier.hpp"
+#include "rosidl_typesupport_introspection_cpp/message_introspection.hpp"
+#include "rosidl_typesupport_introspection_cpp/service_introspection.hpp"
+#include "rosidl_typesupport_introspection_cpp/visibility_control.h"
+
+#include "rosidl_typesupport_introspection_c/field_types.h"
+#include "rosidl_typesupport_introspection_c/identifier.h"
+#include "rosidl_typesupport_introspection_c/message_introspection.h"
+#include "rosidl_typesupport_introspection_c/service_introspection.h"
+#include "rosidl_typesupport_introspection_c/visibility_control.h"
+
+#include "./macros.hpp"
+
+/********** Utility structs **********/
+
+// This struct is used to associate MessageMember the introspection packages
+// with the corresponding MessageMembers struct
+template<typename T>
+struct GenericMembersT;
+
+template<>
+struct GenericMembersT<rosidl_typesupport_introspection_cpp::MessageMember>
+{
+  using type = rosidl_typesupport_introspection_cpp::MessageMembers;
+};
+
+template<>
+struct GenericMembersT<rosidl_typesupport_introspection_c__MessageMember>
+{
+  using type = rosidl_typesupport_introspection_c__MessageMembers;
+};
+
+// This struct is used to associate a C array with a struct that we can metaprogram with
+template<typename T>
+struct GenericCArray;
+
+// multiple definitions of ambiguous primitive types
+SPECIALIZE_GENERIC_C_ARRAY(String, rosidl_generator_c__String)
+SPECIALIZE_GENERIC_C_ARRAY(bool, bool)
+SPECIALIZE_GENERIC_C_ARRAY(byte, uint8_t)
+SPECIALIZE_GENERIC_C_ARRAY(char, char)
+SPECIALIZE_GENERIC_C_ARRAY(float32, float)
+SPECIALIZE_GENERIC_C_ARRAY(float64, double);
+SPECIALIZE_GENERIC_C_ARRAY(int8, int8_t)
+// TODO(jacquelinekay) this results in an ambiguous definition
+// SPECIALIZE_GENERIC_C_ARRAY(uint8, uint8_t)
```

Comment:
I think it's a legacy of ROS 1 that we have both `byte` and `uint8`. The structures and functions are differently named, but aside from that are identical. It might "just work", but if anything were ever different between byte and uint8 in our code, it would be a problem.